pkg/apis/capabilities/v1beta1/product_types.go

```diff
@@ -0,0 +1,1095 @@
+	// headers: As HTTP Headers
+	// query: As query parameters (GET) or body parameters (POST/PUT/DELETE)
+	// authorization: As HTTP Basic Authentication
+	// +optional
+	// +kubebuilder:validation:Enum=headers;query;authorization
+	CredentialsLoc *string `json:"credentials,omitempty"`
+
+	// +optional
+	Security *SecuritySpec `json:"security,omitempty"`
+
+	// TODO GATEWAY RESPONSE
+}
+
+func (a *AppKeyAppIDAuthenticationSpec) SecuritySecretToken() *string {
+	if a.Security == nil {
+		return nil
+	}
+
+	return a.Security.SecuritySecretToken()
+}
+
+func (a *AppKeyAppIDAuthenticationSpec) HostRewrite() *string {
+	if a.Security == nil {
+		return nil
+	}
+
+	return a.Security.HostRewrite()
+}
+
+func (a *AppKeyAppIDAuthenticationSpec) CredentialsLocation() *string {
+	return a.CredentialsLoc
+}
+
+func (a *AppKeyAppIDAuthenticationSpec) AuthAppID() *string {
+	return a.AppID
+}
+
+func (a *AppKeyAppIDAuthenticationSpec) AuthAppKey() *string {
+	return a.AppKey
+}
+
+// UserKeyAuthenticationSpec defines the desired state of User Key Authentication
+type UserKeyAuthenticationSpec struct {
+	// +optional
+	Key *string `json:"authUserKey,omitempty"`
+
+	// Credentials Location available options:
+	// headers: As HTTP Headers
+	// query: As query parameters (GET) or body parameters (POST/PUT/DELETE)
+	// authorization: As HTTP Basic Authentication
+	// +optional
+	// +kubebuilder:validation:Enum=headers;query;authorization
+	CredentialsLoc *string `json:"credentials,omitempty"`
+
+	// +optional
+	Security *SecuritySpec `json:"security,omitempty"`
+
+	// TODO GATEWAY RESPONSE
+}
+
+func (u *UserKeyAuthenticationSpec) SecuritySecretToken() *string {
+	if u.Security == nil {
+		return nil
+	}
+
+	return u.Security.SecuritySecretToken()
+}
+
+func (u *UserKeyAuthenticationSpec) HostRewrite() *string {
+	if u.Security == nil {
+		return nil
+	}
+
+	return u.Security.HostRewrite()
+}
+
+func (u *UserKeyAuthenticationSpec) CredentialsLocation() *string {
+	return u.CredentialsLoc
+}
+
+func (u *UserKeyAuthenticationSpec) AuthUserKey() *string {
+	return u.Key
+}
+
+// AuthenticationSpec defines the desired state of Product Authentication
+type AuthenticationSpec struct {
+	// +optional
+	UserKeyAuthentication *UserKeyAuthenticationSpec `json:"userkey,omitempty"`
+	// +optional
+	AppKeyAppIDAuthentication *AppKeyAppIDAuthenticationSpec `json:"appKeyAppID,omitempty"`
+
+	// TODO OpenID
+}
+
+func (a *AuthenticationSpec) AuthenticationMode() string {
+	// authentication is oneOf by CRD openapiV3 validation
+	if a.UserKeyAuthentication != nil {
+		return "1"
+	}
+
+	// must be appKey&appID
+	return "2"
+}
+
+func (a *AuthenticationSpec) SecuritySecretToken() *string {
+	// authentication is oneOf by CRD openapiV3 validation
+	if a.UserKeyAuthentication != nil {
+		return a.UserKeyAuthentication.SecuritySecretToken()
+	}
+
+	if a.AppKeyAppIDAuthentication == nil {
+		panic("product authenticationspec: both userkey and appid_appkeyare nil")
```

Comment:
Forget about my previous comment.

It seems, as you specified in comment code, that OpenAPI v3 validation is oneOf. I guess you added it manually on the CRD file. This will be hard to maintain and error-prone due to the generator will undo the manual changes but it should work. It'll be really easy to introduce the undoing of changes when we apply big changes to the CRDs

Reply:
If the CRD has openAPI v3 validation, how can the user violate that?

The CRD validation states:
```
oneOf:                          
 - properties:                  
     userkey:                   
       type: object             
   required: ["userkey"]        
 - properties:                  
     appKeyAppID:               
       type: object             
   required: ["appKeyAppID"]    
```

If you try to set none of the auth keys, you will get:
```
validation failure list:
"spec.deployment.apicastSelfManaged.authentication" must validate one and only one schema (oneOf). Found none valid
spec.deployment.apicastSelfManaged.authentication.userkey in body is required
```

If you try to set both auth keys, you will get:

```
validation failure list:
"spec.deployment.apicastSelfManaged.authentication" must validate one and only one schema (oneOf). Found 2 valid alternatives
```

How can you override CRD validation?

Reply:
See my previous comment

Reply:
> This will be hard to maintain and error-prone due to the generator will undo the manual changes but it should work. It'll be really easy to introduce the undoing of changes when we apply big changes to the CRDs

Hopefully this validation will be supported by operator-sdk or underlaying kubebuilder tool https://github.com/operator-framework/operator-sdk/issues/1844

Any validation supported at CRD level, the less code we have to introduce in the operator.

Reply:
:+1: let's hope it's introduced by operator-sdk/kubebuilder tool soon : (